great_expectations/data_context/config_validator/yaml_config_validator.py

```diff
@@ -338,26 +333,16 @@ def _prepare_config_string_with_substituted_variables(
         self, yaml_config: str, runtime_environment: dict, usage_stats_event_name: str
     ) -> str:
         try:
-            substituted_config_variables: Union[
-                DataContextConfig, dict
-            ] = substitute_all_config_variables(
-                self.config_variables,
-                dict(os.environ),
-            )
+            config_provider = self._data_context.config_provider
+            config_values = config_provider.get_values()
 
-            substitutions: dict = {
-                **substituted_config_variables,  # type: ignore[list-item]
-                **dict(os.environ),
-                **runtime_environment,
-            }
+            # While normally we'd just call `self.config_provider.substitute_config()`,
+            # we need to account for `runtime_environment` values that may have been passed.
+            config_values.update(runtime_environment)
```

Comment:
This permanently changes `config_values` which is state inside the `config_provider`.  Is this expected by the caller? This ultimately gets triggered in the public function `test_yaml_config` which doesn't sound like something that mutates state.

Reply:
so everytime I use `config_provider`, the config values are generated JIT - this is due to the fact that a user can modify env vars, config variables, etc during a runtime and we need to always be ready to capture those changes.

This `update()` call modifies the state of the `config_values` dict but that dict is not saved as state on the provider.